README.md
# [RunMany Syntax Highlighting](https://marketplace.visualstudio.com/items?itemName=discretegames.runmany)

[RunMany](https://pypi.org/project/runmany/) is a tool that lets you write and run programs in multiple languages
from one file. This extension provides quick ways to run RunMany files (file extension `.many` or `.runmany`) in VSCode, as well as syntax highlighting for RunMany files and 42
[embeddable programming languages](https://github.com/discretegames/runmanyext/blob/main/supported-languages.csv):

> Ada,
> ARM,
> Bash,
> Batch,
> C,
> C#,
> Clojure,
> COBOL,
> C++,
> Dart,
> Erlang,
> Fortran,
> F#,
> Go,
> Groovy,
> Haskell,
> Java,
> JavaScript,
> Julia,
> Kotlin,
> Lisp,
> Lua,
> MIPS,
> Objective-C,
> OCaml,
> Pascal,
> Perl,
> PHP,
> PowerShell,
> Python,
> Python 2,
> R,
> Racket,
> Ruby,
> Rust,
> Scala,
> Scheme,
> Swift,
> TypeScript,
> x86,
> VBScript, and
> Visual Basic

Languages not [built into VSCode](https://github.com/microsoft/vscode/tree/main/extensions) such as Ada and Pascal
require their own respective VSCode extensions to be installed for their syntax highlighting to work in RunMany files.

RunMany syntax highlighting in action on a
[sample file](https://github.com/discretegames/runmanyext/blob/main/exampleWorkspace/sample.many):

![syntax highlighting sample](https://raw.githubusercontent.com/discretegames/runmanyext/main/images/sample.png)

## Usage

This extension is intended to be used alongside the [RunMany Python package](https://pypi.org/project/runmany/), which is the tool that actually runs `.many` files.

Install the package for [Python 3.6 and above](https://www.python.org/downloads/) with

```text
pip install runmany
```

in terminal, or [see more installation information here](https://github.com/discretegames/runmany#installation).

Then simply hit F5 (or your normal run/debug keybind) in VSCode to run the currently focused RunMany
file in an integrated terminal.

There are additional run buttons on the title menu and command palette.
You can also add a RunMany launch configuration to the [.vscode/launch.json](https://code.visualstudio.com/docs/editor/debugging#_launch-configurations) file of your workspace. For example, this would go in the `"configurations"` array:

```json
{
    "name": "RunMany: Run Current File",
    "type": "runmany",
    "request": "launch",
    "program": "${file}",
    "settingsFile": "",
    "outputFile": ""
}
```

See [here for a full launch.json example](https://github.com/discretegames/runmanyext/blob/main/exampleWorkspace/.vscode/launch.json)
 in an
[example workspace](https://github.com/discretegames/runmanyext/tree/main/exampleWorkspace).

The optional properties `"settingsFile"` and `"outputFile"` correspond to the command line arguments for the
settings JSON and output redirect of the RunMany Python package.

Learn more about the RunMany Python package, including RunMany syntax and what can go in the settings JSON,
on [PyPI](https://pypi.org/project/runmany/)
or [GitHub](https://github.com/discretegames/runmany), or look at some
[example files](https://github.com/discretegames/runmany/tree/main/examples).

## Limitations

Until [Language Server Protocol 3.17](
https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/)
is released there is unfortunately
[no reasonable way](https://code.visualstudio.com/api/language-extensions/embedded-languages#conclusion)
to have underlined errors and warnings (knows as diagnostics) show up in embedded code in VSCode.
Putting the work into support things like autocompletion or semantic highlighting for embedded languages
does not seem worth it without being able to implement diagnostics at the same time,
since they would be the most useful feature.

So for for now, this extension only provides syntax highlighting and easy ways to run RunMany files,
no diagnostics or other features.

## Resources

- [Change Log](https://marketplace.visualstudio.com/items/discretegames.runmany/changelog)

- [GitHub Repo](https://github.com/discretegames/runmanyext)

- [Marketplace Page](https://marketplace.visualstudio.com/items?itemName=discretegames.runmany)

- [RunMany Python Package on PyPI](https://pypi.org/project/runmany/)

- [RunMany Python Package on GitHub](https://github.com/discretegames/runmany)

- [Example RunMany Workspace](https://github.com/discretegames/runmanyext/tree/main/exampleWorkspace)


CHANGELOG.md
# Change Log

All notable changes to the
["RunMany" VSCode extension](https://marketplace.visualstudio.com/items?itemName=discretegames.runmany)
will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.0.0] - 2021-12-03

Updated extension to work with all the new syntax of the [RunMany Python package](https://pypi.org/project/runmany/)
version [2.1.0](https://pypi.org/project/runmany/2.1.0/) and above.

This includes:

- `%%` comment syntax (replacing `%` comments)

- `START:` and `STOP.` syntax (replacing `Exit.`)

- `Settings:` section syntax for embedded JSONs

- `End.` syntax to end sections

- `!!` disabled section and `!` disabled snippet syntax

- `@@` solo section and `@` solo snippet syntax

- Better syntax highlighting and styling overall, removing the need for custom theming

Additionally, these languages have been given embedded syntax highlighting support:

> ARM, Bash, Batch, Erlang, MIPS, OCaml, PowerShell, x86

## [1.0.2] - 2021-10-19

Removed misleading `RunMany: Current File` launch configuration as VSCode auto-removes the `${file}` from it.

## [1.0.1] - 2021-10-09

Re-added bracket and quote auto-completion for `()` `[]` `{}` `''` `""`.

## [1.0.0] - 2021-09-27

Bumped to 1.0.0 because the project is done for the present.

## [0.0.6] - 2021-09-27

- Added ways to run RunMany files directly in VSCode:
  - By hitting F5 (or run/debug keybind) to run current file directly.
  - On command palette and in title menu.
  - Via launch.json configurations that may optionally have `"settingsFile"` and `"outputFile"`.
- Updated readme describing ways to run.

## [0.0.5] - 2021-09-23

- Fixed issue where comments with characters like quotes or colon would appear to be part of code.
- Tweaked sample file and image.

## [0.0.4] - 2021-09-22

- Fixed readme image link issue that only showed up on marketplace and not github.

## [0.0.3] - 2021-09-22

- Improved readme.
- Added `sample.many` sample file with images.
- Improved icon.

## [0.0.2] - 2021-09-21

- Removed undesired icon theme prompt.
- Improved branding.

## [0.0.1] - 2021-09-21

Initial release.

- Readme.
- RunMany syntax highlighting grammar.
- Syntax highlighting for 34 embedded languages.
- Icon for extension and .many files.


testproject/test.c



#include <stdio.h>
int main()
{
	printf("Hello, World!\n");
	return 0;
}

testproject/test.js


function foo():

testproject/test.py
# List and examples of the default VSCode "editor.tokenColorCustomizations" colors (in dark mode at least).
# This example is in Python but the same colors apply to any language.
# Save as a Python (.py) file in VSCode with Dark+ (default dark) color theme to view properly.

    from __future__ import annotations

# #. Key: textMateRules scope
# 1. comments: 'comment'
# 2. strings: 'string'
# 3. keywords: 'keyword'
# 4. numbers: 'constant.numeric'
# 5. types: 'entity.name.type'
# 6. functions: 'entity.name.function'
# 7. variables: 'variable'
# 8. n/a: 'constant.language' (does not autocomplete for some reason)

# See also:
# https://macromates.com/manual/en/language_grammars#naming-conventions
# https://stackoverflow.com/a/45637405
# https://cdn.discordapp.com/attachments/470168395770363905/887061550454693938/unknown.png
# https://github.com/microsoft/vscode/blob/main/extensions/theme-defaults/themes/dark_vs.json


# 1. Comments (comments: 'comment'), inline or block, are green - #6A9955

# this is a comment


# 2. Strings (strings: 'string') are rusty orange/red - #CE9178

'this is a string'
"""this is a string"""
'''this is a string'''
f'this is a string'
r'this is a string'


# 3. Keywords (keywords: 'keyword') are purple - #C586C0

# while and pass are keywords (https://www.programiz.com/python-programming/keyword-list)
while 0:
    pass


# 4. Numbers (numbers: 'constant.numeric'), int or float, are faded light green - #B5CEA8

1234567890
3.14
12e-9
-17
1.8j


# 5. Types (types: 'entity.name.type') are turquoise - #4EC9B0

int
str
float


class my_custom_type:
    pass


# These feel like functions (https://docs.python.org/3/library/functions.html) but are actually types:
zip
map
super

# 6. Functions (functions: 'entity.name.function') are light yellow/wheat - #DCDCAA


print()

id
abs
bin
input
def this_is_a_function(): pass

# 7. Variables (variables: 'variable') are light blue - #9CDCFE


this_is_a_variable = 0

# 8. Constants (n/a: 'constant.language') are blue - #569CD6

None
True
False


class _:
    def _(): pass

# There may be other common colors this list is missing.


out/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DummyDebugSession = exports.activate = void 0;
const vscode = require("vscode");
const vscode_debugadapter_1 = require("vscode-debugadapter");
let runmanyTerminal;
function getTerminal() {
    if (!runmanyTerminal || runmanyTerminal.exitStatus) {
        runmanyTerminal = vscode.window.createTerminal("RunMany");
    }
    return runmanyTerminal;
}
function activate(context) {
    context.subscriptions.push(vscode.debug.registerDebugConfigurationProvider("runmany", new RunManyConfigurationProvider()), vscode.debug.registerDebugAdapterDescriptorFactory("runmany", new InlineDebugAdapterFactory()), vscode.commands.registerCommand("extension.runmany.run", (resource) => {
        if (!resource && vscode.window.activeTextEditor) {
            resource = vscode.window.activeTextEditor.document.uri;
        }
        if (resource) {
            vscode.debug.startDebugging(undefined, {
                name: "RunMany: Run Current File",
                type: "runmany",
                request: "launch",
                program: resource.fsPath,
            });
        }
    }));
}
exports.activate = activate;
class RunManyConfigurationProvider {
    resolveDebugConfiguration(_folder, config, _token) {
        if (!config.name && !config.type && !config.request) {
            const editor = vscode.window.activeTextEditor;
            if (editor && editor.document.languageId === "many") {
                config.name = "RunMany: Dynamic Launch";
                config.type = "runmany";
                config.request = "launch";
                config.program = "${file}";
                config.settingsFile = "";
                config.outputFile = "";
            }
        }
        if (!config.program) {
            return vscode.window
                .showInformationMessage(`No program found to run for RunMany launch configuration.`)
                .then((_) => {
                return undefined;
            });
        }
        if (!config.settingsFile) {
            config.settingsFile = "";
        }
        if (!config.outputFile) {
            config.outputFile = "";
        }
        return config;
    }
}
class InlineDebugAdapterFactory {
    createDebugAdapterDescriptor(session) {
        const terminal = getTerminal();
        terminal.show();
        const programArg = `"${session.configuration.program}"`;
        const settingsArg = session.configuration.settingsFile && ` -s "${session.configuration.settingsFile}"`;
        const outputArg = session.configuration.outputFile && ` -o "${session.configuration.outputFile}"`;
        terminal.sendText(`runmany ${programArg}${settingsArg}${outputArg}`);
        return new vscode.DebugAdapterInlineImplementation(new DummyDebugSession());
    }
}
class DummyDebugSession extends vscode_debugadapter_1.DebugSession {
    initializeRequest() {
        this.sendEvent(new vscode_debugadapter_1.TerminatedEvent());
    }
}
exports.DummyDebugSession = DummyDebugSession;
//# sourceMappingURL=extension.js.map

src/extension.ts
import * as vscode from "vscode";
import { WorkspaceFolder, DebugConfiguration, ProviderResult, CancellationToken } from "vscode";
import { DebugSession, TerminatedEvent } from "vscode-debugadapter";

let runmanyTerminal: vscode.Terminal;

function getTerminal(): vscode.Terminal {
	if (!runmanyTerminal || runmanyTerminal.exitStatus) {
		runmanyTerminal = vscode.window.createTerminal("RunMany");
	}
	return runmanyTerminal;
}

export function activate(context: vscode.ExtensionContext) {
	context.subscriptions.push(
		vscode.debug.registerDebugConfigurationProvider("runmany", new RunManyConfigurationProvider()),
		vscode.debug.registerDebugAdapterDescriptorFactory("runmany", new InlineDebugAdapterFactory()),
		vscode.commands.registerCommand("extension.runmany.run", (resource: vscode.Uri) => {
			if (!resource && vscode.window.activeTextEditor) {
				resource = vscode.window.activeTextEditor.document.uri;
			}
			if (resource) {
				vscode.debug.startDebugging(undefined, {
					name: "RunMany: Run Current File",
					type: "runmany",
					request: "launch",
					program: resource.fsPath,
				});
			}
		})
	);
}

class RunManyConfigurationProvider implements vscode.DebugConfigurationProvider {
	resolveDebugConfiguration(
		_folder: WorkspaceFolder | undefined,
		config: DebugConfiguration,
		_token?: CancellationToken
	): ProviderResult<DebugConfiguration> {
		if (!config.name && !config.type && !config.request) {
			const editor = vscode.window.activeTextEditor;
			if (editor && editor.document.languageId === "many") {
				config.name = "RunMany: Dynamic Launch";
				config.type = "runmany";
				config.request = "launch";
				config.program = "${file}";
				config.settingsFile = "";
				config.outputFile = "";
			}
		}

		if (!config.program) {
			return vscode.window
				.showInformationMessage(`No program found to run for RunMany launch configuration.`)
				.then((_) => {
					return undefined;
				});
		}

		if (!config.settingsFile) {
			config.settingsFile = "";
		}

		if (!config.outputFile) {
			config.outputFile = "";
		}

		return config;
	}
}

class InlineDebugAdapterFactory implements vscode.DebugAdapterDescriptorFactory {
	createDebugAdapterDescriptor(session: vscode.DebugSession): ProviderResult<vscode.DebugAdapterDescriptor> {
		const terminal = getTerminal();
		terminal.show();
		const programArg = `"${session.configuration.program}"`;
		const settingsArg = session.configuration.settingsFile && ` -s "${session.configuration.settingsFile}"`;
		const outputArg = session.configuration.outputFile && ` -o "${session.configuration.outputFile}"`;
		terminal.sendText(`runmany ${programArg}${settingsArg}${outputArg}`);
		return new vscode.DebugAdapterInlineImplementation(new DummyDebugSession());
	}
}

export class DummyDebugSession extends DebugSession {
	protected initializeRequest(): void {
		this.sendEvent(new TerminatedEvent());
	}
}


syntaxes_dev/syntax_maker.py
import csv
import json
from pathlib import Path
from itertools import islice

NAMES_INSERT_INDEX = 3
NAME_PLACEHOLDER = '<NAME>'
ID_PLACEHOLDER = '<ID>'
SOURCE_PLACEHOLDER = '<SOURCE>'


def load_languages(languages_csv):
    languages = []
    with open(languages_csv) as file:
        for row in islice(csv.reader(file), 1, None):
            languages.append(tuple(item.strip() for item in row))
    return languages


def make_include_object(id):
    return {'include': f'#{id}-section'}


def make_language_parts(parts_json, name, id, source):
    with open(parts_json) as file:
        content = file.read()
        content = content.replace(NAME_PLACEHOLDER, name)
        content = content.replace(ID_PLACEHOLDER, id)
        content = content.replace(SOURCE_PLACEHOLDER, source)
        return json.loads(content)


def insert_language_parts(syntax, include, parts, index):
    syntax['repository']['enabled-section']['patterns'].insert(index, include)
    for key, value in parts.items():
        syntax['repository'][key] = value


def save_final_json(final_json, syntax):
    final_json.parent.mkdir(parents=True, exist_ok=True)
    with open(final_json, 'w') as file:
        json.dump(syntax, file, indent=4)


def make_syntax(languages_csv, dev_syntax, parts_json, final_json):
    languages = load_languages(languages_csv)
    for index, (name, id, source) in enumerate(languages, NAMES_INSERT_INDEX):
        include = make_include_object(id)
        parts = make_language_parts(parts_json, name, id, source)
        insert_language_parts(dev_syntax, include, parts, index)
    save_final_json(final_json, dev_syntax)


def load_dev_syntax(dev_json, keys_to_remove, index_to_remove=NAMES_INSERT_INDEX):
    with open(dev_json) as file:
        dev_json = json.load(file)
    del dev_json['repository']['enabled-section']['patterns'][index_to_remove]
    for key in keys_to_remove:
        del dev_json['repository'][key]
    return dev_json


if __name__ == '__main__':
    path = Path(__file__)
    languages_csv = path.parent.with_name('supported-languages.csv')
    dev_json = path.with_name('many.tmLanguage.json')
    dev_syntax = load_dev_syntax(dev_json, ('ada-section', 'ada-snippet'))
    parts_json = path.with_name('language_parts.json')
    final_json = path.parent.parent / 'syntaxes' / 'many.tmLanguage.json'
    make_syntax(languages_csv, dev_syntax, parts_json, final_json)


defaultsyntaxcolors.py
# List and examples of the default VSCode "editor.tokenColorCustomizations" colors (in dark mode at least).
# This example is in Python but the same colors apply to any language.
# Save as a Python (.py) file in VSCode with Dark+ (default dark) color theme to view properly.

# #. Key: textMateRules scope
# 1. comments: 'comment'
# 2. strings: 'string'
# 3. keywords: 'keyword'
# 4. numbers: 'constant.numeric'
# 5. types: 'entity.name.type'
# 6. functions: 'entity.name.function'
# 7. variables: 'variable'
# 8. n/a: 'constant.language' (does not autocomplete for some reason)

# See also:
# https://stackoverflow.com/a/45637405
# https://cdn.discordapp.com/attachments/470168395770363905/887061550454693938/unknown.png
# https://github.com/microsoft/vscode/blob/main/extensions/theme-defaults/themes/dark_vs.json


# 1. Comments (comments: 'comment'), inline or block, are green - #6A9955

# this is a comment


# 2. Strings (strings: 'string') are rusty orange/red - #CE9178

'this is a string'
"""this is a string"""
'''this is a string'''
f'this is a string'
r'this is a string'


# 3. Keywords (keywords: 'keyword') are purple - #C586C0

# while and pass are keywords (https://www.programiz.com/python-programming/keyword-list)
while 0:
    pass

# 4. Numbers (numbers: 'constant.numeric'), int or float, are faded light green - #B5CEA8

1234567890
3.14
12e-9
-17
1.8j


# 5. Types (types: 'entity.name.type') are turquoise - #4EC9B0

int
str
float


class my_custom_type:
    pass


# These feel like functions (https://docs.python.org/3/library/functions.html) but are actually types:
zip
map
super

# 6. Functions (functions: 'entity.name.function') are light yellow/wheat - #DCDCAA

id
abs
bin
input
def this_is_a_function(): pass

# 7. Variables (variables: 'variable') are light blue - #9CDCFE


this_is_a_variable = 0

# 8. Constants (n/a: 'constant.language') are blue - #569CD6

None
True
False


class _:
    def _(): pass

# There may be other common colors this list is missing.
